opcua.cpp

```diff
@@ -1503,6 +1602,253 @@ SOPC_ReturnStatus OPCUA::createS2Subscription()
+ *
+ * @param nodeIdStr          The node ID string of the node to read.
+ * @param outBuiltinTypeId   Pointer to store the BuiltinId of the node's value.
+ * @param outArrayType       Pointer to store the ArrayType of the node's value.
+ * @return                   True if the read operation is successful, false otherwise.
+ */
+bool OPCUA::read(const char *nodeIdStr, SOPC_BuiltinId *outBuiltinTypeId, SOPC_VariantArrayType *outArrayType)
+{
+	if (nodeIdStr == NULL || outBuiltinTypeId == NULL || outArrayType == NULL)
+	{
+		Logger::getLogger()->error("Invalid input parameters for read operation.");
+		return false;
+	}
+
+	SOPC_ReturnStatus status = SOPC_STATUS_NOK;
+	OpcUa_ReadRequest *readRequest = SOPC_ReadRequest_Create(1, OpcUa_TimestampsToReturn_Both);
+	OpcUa_ReadResponse *readResponse = NULL;
+
+	if (readRequest == NULL)
+	{
+		Logger::getLogger()->error("Failed to create ReadRequest.");
+		return false;
+	}
+
+	status = SOPC_ReadRequest_SetReadValueFromStrings(readRequest, 0, nodeIdStr, SOPC_AttributeId_Value, NULL);
+	if (status != SOPC_STATUS_OK)
+	{
+		Logger::getLogger()->error("Failed to set read value for node %s.", nodeIdStr);
+		SOPC_Encodeable_Delete(readRequest->encodeableType, (void **)&readRequest);
+		return false;
+	}
+
+	status = SOPC_ClientHelperNew_ServiceSync(m_connection, readRequest, (void **)&readResponse);
+	SOPC_Encodeable_Delete(readRequest->encodeableType, (void **)&readRequest);
+
+	if (status != SOPC_STATUS_OK || readResponse == NULL)
+	{
+		Logger::getLogger()->error("Read service failed for node %s.", nodeIdStr);
+		return false;
+	}
+
+	if (SOPC_IsGoodStatus(readResponse->ResponseHeader.ServiceResult) &&
+		readResponse->NoOfResults == 1 &&
+		SOPC_IsGoodStatus(readResponse->Results[0].Status))
+	{
+		*outBuiltinTypeId = readResponse->Results[0].Value.BuiltInTypeId;
+		*outArrayType = readResponse->Results[0].Value.ArrayType;
+	}
+	else
+	{
+		Logger::getLogger()->error("Failed to read node %s, StatusCode: 0x%08" PRIX32, nodeIdStr, readResponse->Results[0].Status);
+		status = SOPC_STATUS_NOK;
+	}
+
+	SOPC_Encodeable_Delete(readResponse->encodeableType, (void **)&readResponse);
+	return status == SOPC_STATUS_OK;
+}
+
+/**
+ * Writes a value to a specific node.
+ *
+ * @param nodeIdStr  The node ID string of the node to write to.
+ * @param valueStr   The value to write as a string.
+ * @return           True if the write operation is successful, false otherwise.
+ */
+bool OPCUA::write(const std::string &nodeIdStr, const std::string &valueStr)
+{
+	if (nodeIdStr.empty() || valueStr.empty())
+	{
+		Logger::getLogger()->error("Node ID or value string is empty.");
+		return false;
+	}
+
+	Logger::getLogger()->info("Write request for node %s with value %s", nodeIdStr.c_str(), valueStr.c_str());
+
+	if (m_allowedControlNodes.count(nodeIdStr) == 0)
+	{
+		Logger::getLogger()->error("Write operation not allowed for node %s.", nodeIdStr.c_str());
```

Comment:
It would be worth being explicit that it is disallowed by not being in the control map rathe than the OPCUA is disallowing it, which is unclear in this current message.